File: ts/step1-naming/badRacer.ts
export class Racer {

    private static readonly MOVE_THRESHOLD = 4;
    private static readonly MAX_NAME_LENGTH = 5;


    public canMove(randomValue: number): boolean {
        return randomValue >= Racer.MOVE_THRESHOLD;
    }

    public isValidName(name: string | null): boolean {
        if (name === null) return false;
        if (name.length === 0) return false;
        if (name.length > Racer.MAX_NAME_LENGTH) return false;
        return true;
    }


    public findWinners(distanceByRacer: Map<string, number>): string[] {
        let maxDistance = 0;

        for (const distance of distanceByRacer.values()) {
            if (distance > maxDistance) {
                maxDistance = distance;
            }
        }

        const winners: string[] = [];

        for (const [name, distance] of distanceByRacer) {
            if (distance === maxDistance) {
                winners.push(name);
            }
        }

        return winners;
    }
}

File: java/step1-naming/src/main/java/badcode/BadRacer.java
package badcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class BadRacer {

    private static final int MIN_MOVE_VALUE = 4;
    private static final int MAX_NAME_LENGTH = 5;
    private static final int INIT_DISTANCE = 0;

    public boolean canMove(int randomValue) {
        return randomValue >= MIN_MOVE_VALUE;
    }

    public boolean isValidName(String name) {
        if (name == null) {
            return false;
        }
        if (name.length() > MAX_NAME_LENGTH) {
            return false;
        }
        if (name.isEmpty()) {
            return false;
        }
        return true;
    }

    public List<String> getWinners(Map<String, Integer> car) {
        int maxDistance = INIT_DISTANCE;
        for (Map.Entry<String, Integer> e : car.entrySet()) {
            if (e.getValue() > maxDistance) {
                maxDistance = e.getValue();
            }
        }
        List<String> winners = new ArrayList<>();
        for (Map.Entry<String, Integer> e : car.entrySet()) {
            if (e.getValue() == maxDistance) {
                winners.add(e.getKey());
            }
        }
        return winners;
    }
}
